pkg/scan/docker/docker.go

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) 2021 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package docker
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	dockertypes "github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/google/go-containerregistry/pkg/crane"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/daemon"
+	"github.com/mholt/archiver/v3"
+	"github.com/palantir/log4j-sniffer/pkg/archive"
+	"github.com/palantir/log4j-sniffer/pkg/crawl"
+	"github.com/palantir/log4j-sniffer/pkg/scan"
+	"github.com/pkg/errors"
+)
+
+type Scanner struct {
+	config     scan.Config
+	crawler    crawl.Crawler
+	reporter   *crawl.Reporter
+	identifier crawl.Identifier
+	client     client.CommonAPIClient
+}
+
+func NewDockerScanner(config scan.Config, stdout, stderr io.Writer) (*Scanner, error) {
+	dockerClient, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create docker client")
+	}
+
+	return &Scanner{
+		config: config,
+		crawler: crawl.Crawler{
+			ErrorWriter: stderr,
+			IgnoreDirs:  config.Ignores,
+		},
+		reporter: &crawl.Reporter{
+			OutputJSON:      config.OutputJSON,
+			OutputWriter:    stdout,
+			DisableCVE45105: config.DisableCVE45105,
+		},
+		identifier: crawl.NewIdentifier(config.ArchiveListTimeout, archive.WalkZipFiles, archive.WalkTarGzFiles),
+		client:     dockerClient,
+	}, nil
+}
+
+func (d Scanner) ScanImages(ctx context.Context) (int64, error) {
+	imageList, err := d.client.ImageList(ctx, dockertypes.ImageListOptions{})
+	if err != nil {
+		return 0, errors.Wrap(err, "could not list docker images")
+	}
+
+	var stats crawl.Stats
+	for _, image := range imageList {
+		if len(image.RepoTags) == 0 {
+			continue
+		}
+		d.reporter.SetImageID(image.ID)
+		d.reporter.SetImageTags(image.RepoTags)
```

Comment:
I'm a little confused about why we overwrite the image id and tags every image but the count will be the sum of across all images.

Would it make sense to have a separate reporter for docker? I'm not sure at this point.

Reply:
I think it's a design/usability question, do we want a per image summary or a summary of all the files and findings found in all images. I'll look into a separate reporter it might make things easier